smart-contracts/contracts/Blocklist.sol

```diff
@@ -0,0 +1,136 @@
+pragma solidity 0.5.16;
+
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+
+/**
+ * @title Blocklist
+ * @dev This contract manages a list of addresses and has a simple CRUD
+ */
+contract Blocklist is Ownable {
+  /**
+   * @dev The index of each user in the list
+   */
+  mapping(address => uint256) private _userIndex;
+
+  /**
+   * @dev The list itself
+   */
+  address[] private _userList;
+
+  /**
+   * @notice Event emitted when a user is added to the blocklist
+   */
+  event addedToBlocklist(address indexed account, address by);
+
+  /**
+   * @notice Event emitted when a user is removed from the blocklist
+   */
+  event removedFromBlocklist(address indexed account, address by);
+
+  /**
+   * @notice Modifier to facilitate checking the blocklist
+   */
+  modifier onlyInBlocklist(address account) {
+    require(isBlocklisted(account), "Not in blocklist");
+    _;
+  }
+
+  /**
+   * @notice Modifier to facilitate checking the blocklist
+   */
+  modifier onlyNotInBlocklist(address account) {
+    require(!isBlocklisted(account), "Already in blocklist");
+    _;
+  }
+
+  /**
+   * @dev Adds an address to the blocklist
+   * @param account The address to add
+   * @dev Fails if the address was already blocklisted
+   */
+  function _addToBlocklist(address account) private onlyNotInBlocklist(account) returns(bool) {
+    _userIndex[account] = _userList.length;
+    _userList.push(account);
+
+    emit addedToBlocklist(account, msg.sender);
+
+    return true;
+  }
+
+  /**
+   * @notice Adds many addresses to the blocklist at once
+   * @param accounts[] The list of addresses to add
+   * @dev Fails if at least one of the addresses was already blocklisted
+   */
+  function batchAddToBlocklist(address[] calldata accounts) external onlyOwner {
+    for (uint256 i = 0; i < accounts.length; i++) {
+      require(_addToBlocklist(accounts[i]));
+    }
+  }
+
+  /**
+   * @notice Adds an address to the blocklist
+   * @param account The address to add
+   * @dev Fails if the address was already blocklisted
+   */
+  function addToBlocklist(address account) external onlyOwner returns(bool) {
+    return _addToBlocklist(account);
+  }
+
+  /**
+   * @dev Removes an address from the blocklist
+   * @param account The address to remove
+   * @dev Fails if the address was not blocklisted
+   */
+  function _removeFromBlocklist(address account) private onlyInBlocklist(account) returns(bool) {
```

Comment:
I am suspicious this function does not do what we want it to do, just wanted to mark my concern so we can talk in the morning Daniel.